### DISCUSSION TITLE
Is there any way to detect emojis in text?

### DISCUSSION THREAD

**Discussion by Ender22** [airsdk/Adobe-Runtime-Support#2434] [General]
As far as I know, Emojis can't consistently be displayed on iOS in Air with starling/feathers due to various previously logged issues (you need to use StageTextTextEditor but even then the text glitches often and gets vertically cut off) 

So I'm wondering if there is any way to detect them in a textfield so that I can apply various workarounds? 

Regex's seem like they can't work because AS3 does not support the /u option for comparing unicode values. 

I've also tried the following:
`var emojiCodePoints:Array = ["😃", "😀", "😄", "😁"];
var emojiCodePoints2:Array = ["U+1F600", "U+1F603", "U+1F604", "U+1F601"];
			
			for (var i:int = 0; i < text.length; i++) 
			{
				var codePoint:uint = text.charCodeAt(i);
				if (emojiCodePoints.indexOf(codePoint) != -1)
					return true;
			}`
but it seems like this doesn't work as it always returns false (with either type of array)

wondering if anyone has any hidden tricks.

**Comment by ajwfrost** [comment]
>  AS3 does not support the /u option for comparing unicode values.

Well the regexp thing should be able to support this! but doesn't... (I was looking at [this link for a JS equivalent](https://melvingeorge.me/blog/check-if-string-contain-emojis-javascript))

So until we fix that, you could do it the slow way like this:
```
			for (var i:int = 0; i < str.length; i++) 
			{
				var codePoint:uint = str.charCodeAt(i);
				if (codePoint == 0x00A9 || codePoint == 0x00AE || (codePoint >= 0x2000 && codePoint <= 0x3300)) trace("MATCH");
				else if (codePoint == 0x0D83C || codePoint == 0xD83D || codePoint == 0xD83E)
				{
					codePoint = str.charCodeAt(i + 1);
					if (codePoint >= 0xD000 && codePoint <= 0xDFFF) trace("MATCH");
				}
			}
```

Internally, everything seems to be properly converted into UTF-8, and the regular expression library is built with unicode support, so I'm not quite sure why it doesn't actually match it... we can look at this!
```
			var regExp : RegExp = /(\u00a9|\u00ae|[\u2000-\u3300]|\ud83c[\ud000-\udfff]|\ud83d[\ud000-\udfff]|\ud83e[\ud000-\udfff])/;
			var matches : Array = str.match(regExp); // returns null due to internal failure..
```

**Reply by pnico** [reply]
I think this part `\u2000-\u3300` will actually match some non-emoji chars. changing your last 3 options to just write them out the verbose way works for me for the emoji chars though. maybe something with splitting out the 'surrogate' part in the regex? 
```
var regExp : RegExp = /(\u00a9|\u00ae|[\u2000-\u3300]|[\ud83c\ud000-\ud83c\udfff]|[\ud83d\ud000-\ud83d\udfff]|[\ud83e\ud000-\ud83e\udfff])/;
```

**Reply by ajwfrost** [reply]
That regex was something I got off the web, I did wonder as I thought all the emojis were done with surrogate pairs.. 

But thank you for this -> we had got as far as working out that the format of `\ud83c[\ud000-\udfff]` was causing a problem as this is writing in the first code of a surrogate pair but the next character is `[` which isn't valid.

Writing them as `[\ud83c\ud000-\ud83c\udfff]` turns it into a valid character range, with each character being made up of the two surrogate codes... works for me!

**Comment by Ender22** [comment]
Cool glad to hear you guys will look into the regex, and thanks for the example, I didn't think to try it that way!

**Reply by ajwfrost** [reply]
@Ender22 are you able to try the regex suggested above by @pnico ? My code now:
```
			var str : String = "Hello 😃";
			var regExp : RegExp = /(\u00a9|\u00ae|[\u2000-\u3300]|[\ud83c\ud000-\ud83c\udfff]|[\ud83d\ud000-\ud83d\udfff]|[\ud83e\ud000-\ud83e\udfff])/;
			var matches : Array = str.match(regExp);
			if (matches == null) trace("REGEXP ERROR");
			else if (matches.length) trace("MATCHES");
			else trace("NO MATCH");
```
displays `MATCHES`

**Reply by Ender22** [reply]
Hmm unfortunately that code is giving some false positives when different languages that my app uses are included in the text. 
For example:
`"How are you, Как дела, คุณสบายดีไหม, Bạn khỏe không, 你好吗, 你好嗎, Сен жақсың, ¿Cómo estás?"` 
is returning as a match.
I've tried removing the \u2000-\u3300 part as well, but it still comes as a match. I'll try to see if I can find a more specific regex

**Reply by Ender22** [reply]
FYI I had some confusing experiments until I realized that if this code does not match an emoji it actually traces REGEXP ERROR

i.e. the following will give REGEXP ERROR
```
var str : String = "Hello";
			var regExp : RegExp = /(\u00a9|\u00ae|[\u2000-\u3300]|[\ud83c\ud000-\ud83c\udfff]|[\ud83d\ud000-\ud83d\udfff]|[\ud83e\ud000-\ud83e\udfff])/;
			var matches : Array = str.match(regExp);
			if (matches == null) trace("REGEXP ERROR");
			else if (matches.length) trace("MATCHES");
			else trace("NO MATCH");
```

**Reply by Ender22** [reply]
It doesn't seem to be a problem with the regex itself but something with as3. If I use: https://www.regextester.com/ with the same regex then my sentence does not cause a match.

![no-match](https://user-images.githubusercontent.com/8116040/213390026-4b9e9392-361d-471c-8c9f-88efbda2b27e.jpg)

and it correctly matches an emoji:

![1 match](https://user-images.githubusercontent.com/8116040/213390090-e23cede5-de45-4ca9-aa52-b00d094cbd23.jpg)

But in as3 it is a match on the '?' character for some reason in the following:

```
var text:String = "How are you, Как дела, คุณสบายดีไหม, Bạn khỏe không, 你好吗, 你好嗎, Сен жақсың, ¿Cómo estás?";
var regExp : RegExp = /(\u00a9|\u00ae|[\u2000-\u3300]|[\ud83c\ud000-\ud83c\udfff]|[\ud83d\ud000-\ud83d\udfff]|[\ud83e\ud000-\ud83e\udfff])/;
var matches : Array = text.match(regExp);
```

I suppose I could remove all "?" before checking for a match, but any idea why?

**Reply by ajwfrost** [reply]
This may be matching the 0xA9/0xAE values - not sure why they are actually included in that regex query string! The 0xAE (174) value looks like it could be the same as is used for the start of a question (it's apparently like a double less-than symbol, I've seen that used instead of the upside-down question mark when switching between different fonts/locales sometimes...)

So yes, perhaps the regex query should just be the `\ud83c\ud000` part onwards..

> if this code does not match an emoji it actually traces REGEXP ERROR

Ah, yes I've just re-read the documentation on `String.match()` and if we don't have the `/g` option at the end then it will return a null result if there's no match. If we add `/g` to the end then it should return the empty array (but probably wasteful to create a new Array object..)

**Reply by Ender22** [reply]
Cool thanks a lot for the help, I'll try removing those values but either way this is working nicely!

**Reply by pnico** [reply]
Yeah there are tons of posts about how to get the perfect regex to detect emojis, it's tough. We use one similar to this ^, ours has false positives but it's good enough for our purpose. also if you just want to check a block of text to see if you need to render it specially for emojis, just `myTextBlock.search(myEmojiDetectingRegex)` on the whole text block without looping through it is enough, if it returns -1 you can use a normal text field and if not.. good luck 😅 we have a lot of issues like you're describing with clipping, empty space etc. when you render the bitmap from a StageText. It's easier if you have a fixed-size area to render to, but it's tricky. So yes, built-in emoji support in AIR would be really nice

**Reply by raresn** [reply]
i agree - we have a lot of trouble with emojis in text!

**Reply by Ender22** [reply]
Yeah currently my strategy for displaying emojis in a non-fixed-sized area is to have delayedcalls that move the textfield up/down and increase/decrease it's height by 0.01 pixels every 0.5 seconds... in general it keeps the text displaying okay but yeah, it's ugly!

Thanks for the search tip!

**Comment by waveofthought-code** [comment]
While not the most ideal situation, in the past I leveraged browser/javascript functionality for this.  Browser (or in this case, WebView) unicode detection is far superior and making an ExternalInterface call to parse emojis actually worked for me.   Its clunky to be sure, but it worked.
